modules/optimization/src/functions/NearestReporterCoordinatesFunction.C

```diff
@@ -0,0 +1,205 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NearestReporterCoordinatesFunction.h"
+
+registerMooseObject("OptimizationApp", NearestReporterCoordinatesFunction);
+
+InputParameters
+NearestReporterCoordinatesFunction::validParams()
+{
+  InputParameters params = OptimizationFunction::validParams();
+  params.addClassDescription(
+      "Function based on the nearest point to coordinates and values defined by a vector of "
+      "values from a vector-postprocessor or reporter, interpolates linearly in time with "
+      "transient data.");
+  params.addParam<ReporterName>("x_coord_name",
+                                "Name of vector-postprocessor or reporter vector containing "
+                                "x-coordinate of points, default is assumed to be all 0s.");
+  params.addParam<ReporterName>("y_coord_name",
+                                "Name of vector-postprocessor or reporter vector containing "
+                                "y-coordinate of points, default is assumed to be all 0s.");
+  params.addParam<ReporterName>("z_coord_name",
+                                "Name of vector-postprocessor or reporter vector containing "
+                                "z-coordinate of points, default is assumed to be all 0s.");
+  params.addParam<ReporterName>("time_name",
+                                "Name of vector-postprocessor or reporter vector containing time, "
+                                "default is assumed to be all 0s.");
+  params.addRequiredParam<ReporterName>(
+      "value_name", "Name of vector-postprocessor or reporter vector containing value data.");
+  return params;
+}
+
+NearestReporterCoordinatesFunction::NearestReporterCoordinatesFunction(
+    const InputParameters & parameters)
+  : OptimizationFunction(parameters),
+    ReporterInterface(this),
+    _coordx(isParamValid("x_coord_name") ? getReporterValue<std::vector<Real>>("x_coord_name")
+                                         : _empty_vec),
+    _coordy(isParamValid("y_coord_name") ? getReporterValue<std::vector<Real>>("y_coord_name")
+                                         : _empty_vec),
+    _coordz(isParamValid("z_coord_name") ? getReporterValue<std::vector<Real>>("z_coord_name")
+                                         : _empty_vec),
+    _coordt(isParamValid("time_name") ? getReporterValue<std::vector<Real>>("time_name")
+                                      : _empty_vec),
+    _values(getReporterValue<std::vector<Real>>("value_name"))
+{
+}
+
+Real
+NearestReporterCoordinatesFunction::value(Real t, const Point & p) const
+{
+  const std::array<std::pair<Real, std::size_t>, 2> tv = findNearestPoint(t, p);
+
+  if (tv[0].second == tv[1].second)
```

Comment:
add a comment to say when this happens (edge of points domain or edge of time domain?)